text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,1088 @@
+
+    randomness = vrf_bytes(32, randomness_vrf_input, randomness_vrf_output);
+
+    accumulator = BLAKE2(32, CONCAT(accumulator, randomness));
+```
+
+The updated `accumulator` value is stored on-chain.
+
+The randomess accumulated during epoch `N` will be used, at the start of the
+next epoch (`N+1`), as an input to compute the `NextEpochDescriptor`
+`randomness` element (see section 6.1). 
+
+As outlined throughout the document, epoch randomness value secures various
+protocol-specific functions, including ticket generation and assignment of
+fallback slots (refer to section 6.4.2). Additionally, users may utilize this
+value for other purposes as needed.
+
+
+## 7. Drawbacks
+
+None
+
+## 8. Testing, Security, and Privacy
+
+The reference implementation for this RFC will be tested on testnets first.
+
+An audit may be required to ensure the implementation does not introduce unwanted side effects
+
+## 9. Performance, Ergonomics, and Compatibility
+
+### 9.1. Performance
+
+The utilization of Sassafras consensus represents a significant advancement in
+the mitigation of short-lived fork occurrences.
+
+Generation of forks are not possible when following the protocol and the only source
+of forks is network partitioning. In this case, on recovery, the decision of
+which fork to follow is not opinionated and there is only one choice.
+
+### 9.2. Ergonomics
+
+No specific considerations.
+
+### 9.3. Compatibility
+
+The adoption of Sassafras impacts native client code and thus can't be
+introduced via a simple runtime upgrade.
+
+A deployment strategy should be carefully engineered for live networks.
+
+This subject is left open for a dedicated RFC.
+
+
+## 10. Prior Art and References
+
+- Web3 Foundation research page: https://research.web3.foundation/Polkadot/protocols/block-production/SASSAFRAS
+- Sassafras whitepaper: https://eprint.iacr.org/2023/031.pdf
+- Ring-VRF whitepaper: https://eprint.iacr.org/2023/002.pdf
+- Sassafras reference implementation tracking issue: https://github.com/paritytech/substrate/issues/11515
+- Sassafras reference implementation main PR: https://github.com/paritytech/substrate/pull/11879
+
+
+## 11. Unresolved Questions
+
+None
+
+
+## 12. Future Directions and Related Material
+
+While this RFC lays the groundwork and outlines the core aspects of the
+protocol, several crucial topics remain to be addressed in future RFCs to ensure
+the protocol's completeness and security.
+
+These topics include:
+
+### 12.1. Interactions with the Runtime
+
+- **Outbound Interface**. Interfaces exposed by the host which are required by the runtime.
+  These are commonly dubbed *Host Functions*.
+
+- **Unrecorded Inboud Interfaces**. Interfaces exposed by the runtime which are required by the host.
+  These are commonly dubbed *Runtime APIs*.
+
+- **Transactional Inboud Interfaces**. Interfaces exposed by the runtime which alter the state.
+  These are commonly dubbed *Extrinsics* and *Inherents*.
```

Comment:
@tomaka given that this RFC already contains quite a lot of information, I thought that may be worth dedicating an RFC to all the topics which concerns the actual integration with Polkadot.

This RFC is more intended to be a bridge between the research and the engineering. And propose the actual structures and procedures to implement the protocol.

I want to propose soon an RFC containing all the interactions which are expected to happen between the runtime, the host and the external world. That is runtime api, host functions, storage entries, extrinsics and inherents.

IMO is worth to keep these topics separated as:
- there is a separation between the protocol description and the actual integration with a system (i.e. Polkadot)
- maybe is better to don't bloat too much this RFC